gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+filter as follows:
+<pre>
+construct_normal_gcs_filter(block, fp) -> []byte:
+    let siphash_key = block.hash()[:16]
+
+    let P = 1 << fp
+
+    let raw_items = []
+    for tx in block.transactions:
+        let txid = tx.hash()
+        raw_items.append(txid)
+
+        for output in tx.outputs:
+            let output_bytes = extract_push_datas(output.script)
+            raw_items.append(output_bytes)
+
+        if tx.is_coinbase():
+            continue
+
+        for input in tx.inputs:
+            // Inputs serialized as they are on the wire in transactions.
+            // Input index serialized in little-endian.
+            let input_bytes = input.hash || input.index
+            raw_items.append(input_bytes)
+
+    let N = len(raw_items)
+    let F = N * P
+
+    let hashed_items = []
+    for raw_item in raw_items:
+        let hashed_item = (siphash_key(siphash_key, raw_item) * F) >> 64
+        hashed_items.append(hashed_item)
+
+    hashed_items.sort()
+
+    gcs_compress(hashed_items, fp)
+</pre>
+
+
+Given a Bitcoin block, a full-node construct an <code>Extended</code> compact filter as follows
+<pre>
+construct_extended_gcs_filter(block, fp) -> []byte:
+
+    let siphash_key = block.hash()[:16]
+
+    let P = 1 << fp
+
+    let raw_items = []
+    for tx in block.transactions:
+        if tx.is_coinbase():
+           continue
+
+        for input in tx.inputs:
+            for wit_elem in input.witness:
+                raw_items.append(wit_elem)
+
+            let sig_script_pushes = extract_push_datas(input.sig_script)
+            for push in sig_script_pushes:
+                raw_items.append(push)
+
+    let N = len(raw_items)
+    let F = N * P
+
+    let hashed_items = []
+    for raw_item in raw_items:
+        let hashed_item = (siphash_key(siphash_key, raw_item) * F) >> 64
+        hashed_items.append(hashed_item)
+
+    hashed_items.sort()
+
+    gcs_compress(hashed_items, fp)
+</pre>
+
+==== Filter Capability Querying ====
+
+As it's feasible that in the future, this document is extended to encompass
+additional filter encoding algorithms or filter contents, we define a new p2p
+message that allows light clients to ascertain which filters a node supports.
+
+The <code>getcftypes</code> message is an ''empty message'' whose command string is:
+<code>getcftypes</code>
+
+A full-node that receives a <code>getcftypes</code> message should respond with a
+<code>cftypes</code> message which is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| Var-Int
+| NumFilters
+| uint64
+| The number of supported filters.
+|-
+| NumFilters
+| SupportedFilters
+| [NumFilterBytes]byte
+| A byte slice with each byte denoting a supported filter type
+|}
+
+
+==== Compact Filter Header Chain ====
+
+As the filters described in this BIP ''are not'' consensus critical, meaning
+each filter is validated by full-nodes and committed into blocks by miners, we
```

Comment:
Wait, huh? I think you mean meaning filters are *not* validated by full-nodes and *not* committed to in blocks.

Reply:
More importantly, dunno why you have to specify this? It is binding insofar as the protocol says it must be X, and light clients will think you're broken if its not X.

Reply:
Ah yeh, nice catch! Modified to say "not". Specifying it to give the reader the rationale to why the header chain exists.